rust/src/wallet.rs

```diff
@@ -1,45 +1,81 @@
 use bdk::electrum_client::Client;
 use bdk::KeychainKind;
 use bdk::SyncOptions;
-use bdk::Wallet;
+use state::Storage;
 use std::str::FromStr;
+use std::sync::Mutex;
 
 pub const MAINNET_ELECTRUM: &str = "ssl://blockstream.info:700";
 pub const TESTNET_ELECTRUM: &str = "ssl://blockstream.info:993";
 
-pub fn init_wallet(network: &str) -> Result<()> {
-    let network = Network::from_str(network)?;
+/// Wallet has to be managed by Rust as generics are not support by frb
+static WALLET: Storage<Mutex<Wallet>> = Storage::new();
+
+pub struct Wallet {
+    blockchain: ElectrumBlockchain,
+    wallet: bdk::Wallet<MemoryDatabase>,
+}
 
-    let electrum_url = match network {
-        Network::Bitcoin => MAINNET_ELECTRUM,
-        Network::Testnet => TESTNET_ELECTRUM,
-        _ => bail!("Only public networks are supported"),
-    };
+impl Wallet {
+    pub fn new(network: &str) -> Result<Wallet> {
+        let network = Network::from_str(network)?;
 
-    let seed = Bip39Seed::new()?;
-    let ext_priv_key = seed.derive_extended_priv_key(Network::Testnet)?;
+        let electrum_url = match network {
+            Network::Bitcoin => MAINNET_ELECTRUM,
+            Network::Testnet => TESTNET_ELECTRUM,
+            _ => bail!("Only public networks are supported"),
+        };
 
-    let client = Client::new(electrum_url)?;
-    let blockchain = ElectrumBlockchain::from(client);
+        let seed = Bip39Seed::new()?;
+        let ext_priv_key = seed.derive_extended_priv_key(Network::Testnet)?;
 
-    let wallet = Wallet::new(
-        bdk::template::Bip84(ext_priv_key, KeychainKind::External),
-        Some(bdk::template::Bip84(ext_priv_key, KeychainKind::Internal)),
-        ext_priv_key.network,
-        MemoryDatabase::new(),
-    )?;
+        let client = Client::new(electrum_url)?;
+        let blockchain = ElectrumBlockchain::from(client);
 
-    wallet.sync(&blockchain, SyncOptions::default())?;
+        let wallet = bdk::Wallet::new(
+            bdk::template::Bip84(ext_priv_key, KeychainKind::External),
+            Some(bdk::template::Bip84(ext_priv_key, KeychainKind::Internal)),
+            ext_priv_key.network,
+            MemoryDatabase::new(),
+        )?;
 
+        Ok(Wallet { blockchain, wallet })
+    }
+
+    pub fn sync(&self) -> Result<bdk::Balance> {
+        self.wallet.sync(&self.blockchain, SyncOptions::default())?;
+
+        let balance = self.wallet.get_balance()?;
+        println!("Wallet balance: {} SAT", &balance);
+        Ok(balance)
+    }
+}
+
+/// Boilerplate wrappers for using Wallet with static functions in the library
+
+pub fn init_wallet(network: &str) -> Result<()> {
+    WALLET.set(Mutex::new(Wallet::new(network)?));
     Ok(())
 }
 
+pub fn get_balance() -> Result<bdk::Balance> {
+    println!("Wallet sync called");
+    WALLET
+        .try_get()
+        .context("Wallet uninitialised")?
+        .lock()
+        .map_err(|_| anyhow!("cannot acquire wallet lock"))?
+        .sync()
```

Comment:
Just out of interest, we don't have to change anything for now:

Would we be able to let rust handle the sync interval given that the wallet lives in rust? 
Then we could just read the latest balance instead of awaiting it. 

This would have the advantage that dart does not have to await the balance - potential disadvantage could be that we have to add a check for the balance to be outdated - but we need logic around that anyway, for the current code I would expect to have a timeout around sync at some point.

Reply:
that would be harder, but not impossible. It would require us to dedicate a thread to run tokio runtime or something and never return - will need to have a deeper look into how to even ensure this thread would not be assigned to other calls from Flutter later etc.
It will require more `Act<Mutex>` to store the value too, as you're not guaranteed that you will get the value (unless you "push" via an async stream, which I'll be trialing later on today). 

pro of the current situation (so far):
- flutter awaits, which is non-blocking, and we're only using one thread from the Rust thread poo
- no need for 2 separate async runtimes (dart + rust) - this might need to change, we'll see.

tl;dr: stay tuned, there might be changes :)